### DISCUSSION TITLE
How to map @property attribute imperatively?

### DISCUSSION THREAD

**Discussion by jalvespinto** [sqlalchemy/sqlalchemy#7574] [Usage Questions] [answered]
How can I map a @property attribute imperatively?
Here is what I was trying to do, but `id` won't be mapped doing this way. I tried using @hybrid_property instead, but nothing changes. Any suggestions? tks
```
from dataclasses import dataclass
import uuid
from sqlalchemy.ext.hybrid import hybrid_property
from sqlalchemy.orm import relationship, registry
from sqlalchemy import Table, Column
from src.vlep.adapters.orm import start_mappers, mapper_registry
from sqlalchemy import create_engine
from sqlalchemy.sql.sqltypes import String

mapper_registry = registry()


@dataclass
class Product:

    _id: str = uuid.uuid4().hex

    @property
    def id(self):
        return self._id


products = Table(
    'products', mapper_registry.metadata,
    Column('id', String(255), primary_key=True, unique=True),
)


def start_mappers():
    product_mapper = mapper_registry.map_imperatively(
        Product,
        products,
        properties={'id': products.c.id} # this will create the attribute, but it will be NoneType. Without this products can not assemble any primary key.
    )


if __name__ == "__main__":
    engine = create_engine("sqlite:///:memory:")
    mapper_registry.metadata.create_all(engine)
    start_mappers()

    p = Product()

    print(type(p.id)) #result is NoneType
```

**Comment by zzzeek** [comment, marked as answer]
you need to have an attribute that is mapped directly to the column.  when you use `@property`, that name used for the function that's decorated by `@property` is the one that would not be mapped; the thing it accesses instead is what's mapped.

based on your example it looks like you want the attribute named "_id" to be mapped.  using imperative mapping that would look like:

```py
   product_mapper = mapper_registry.map_imperatively(
        Product,
        products,
        properties={'_id': products.c.id}
    )
```

there's also a helper called `synonym` that features support for working with explicit properties, that is documented at https://docs.sqlalchemy.org/en/14/orm/mapped_attributes.html#synonyms.  however, when you have an explicit `@property` like that where we would assume you have some kind of custom code inside your "id" method, the "synonym" isn't really getting you anything and you might just keep it at `@property`.  on the other hand, if you want to use the class-level attribute `Product.id` in queries, then you'd switch `@property` for `@hybrid_property`, which adds class-level behaviors to the attribute.

**Reply by jalvespinto** [reply]
Tks man!
